src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp

```diff
@@ -0,0 +1,331 @@
+  // Suppose we keep up to and including the (ax)^(2n) term in the
+  // series.  Then the series is accurate if the (ax)^{2n+2} term (the
+  // next term in the series) is small, i.e. if
+  // (2n+2)(ax)^{2n+2}/(2n+3)! < eps.
+  //
+  // For the worst case of (2n+2)(ax)^{2n+2}/(2n+3)! == eps, the direct
+  // evaluation still has a relative error of 3*eps/(ax)^2, which evaluates to
+  // error = 3*eps* eps^{-1/(n+1)} * ((2n+2)/(2n+3)!)^{1/(n+1)}.
+  // This can be rewritten as
+  // error = 3 * [eps^n*((2n+2)/(2n+3)!)]^{1/(n+1)}.
+  //
+  // For certain values of n:
+  // n=1    error=3*sqrt(eps/30)               ~ 5e-9
+  // n=2    error=3*(eps^2/840)^(1/3)          ~ 7e-12
+  // n=3    error=3*(eps^3/45360)^(1/4)        ~ 2e-13
+  // n=4    error=3*(eps^4/3991680)^(1/5)      ~ 2e-14
+  // n=5    error=3*(eps^5/518918400)^(1/6)    ~ 5e-15
+  // n=6    error=3*(eps^6/93405312000)^(1/7)  ~ 1e-15
+  //
+  // We gain less and less with each order.
+  //
+  // So here we choose n=3.
+  // Then the series above can be rewritten
+  // d/d(log x) [ sin(ax)/x ] = a/x^2 [- 2(ax)^2/3! + 4(ax)^4/5! - 6(ax)^6/7!]
+  //                          = -a^3/3 [ 1 - 4*3*(ax)^2/5! + 6*3*(ax)^4/7!]
+  const double ax = a*x;
+  return pow<8>(ax) < 45360.0 * std::numeric_limits<double>::epsilon()
+             ? (-cube(a) / 3.0) *
+                   (1.0 + square(ax) * (-0.1 + square(ax) / 280.0))
+             : (a * cos(ax) - sin(ax) / x) / square(x);
+}
+DataVector dlogx_sin_ax_over_x(const DataVector& x, double a) noexcept {
+  DataVector result(x);
+  for (size_t i = 0; i < x.size(); ++i) {
+    result[i] = dlogx_sin_ax_over_x(x[i], a);
+  }
+  return result;
+}
+}  // namespace Endcap_detail
+
+Endcap::Endcap(const std::array<double, 3>& center, double radius,
+               double z_plane) noexcept
+    : center_(center),
+      radius_([&]() noexcept {
+        // The equal_within_roundoff below has an implicit scale of 1,
+        // so the ASSERT may trigger in the case where we really
+        // want an entire domain that is very small.
+        ASSERT(not equal_within_roundoff(radius, 0.0),
+               "Cannot have zero radius");
+        return radius;
+      }()),
+      theta_max_([&]() noexcept {
+        const double cos_theta_max = (z_plane - center_[2]) / radius_;
+        ASSERT(abs(cos_theta_max) < 1.0,
+               "Plane must intersect sphere, and at more than one point");
+        return acos(cos_theta_max);
+      }()) {}
+
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> Endcap::operator()(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  const ReturnType& xbar = source_coords[0];
+  const ReturnType& ybar = source_coords[1];
+  const ReturnType rhobar = sqrt(square(xbar) + square(ybar));
+  const ReturnType sin_factor =
+      radius_ * Endcap_detail::sin_ax_over_x(rhobar, theta_max_);
+  ReturnType z = radius_ * cos(rhobar * theta_max_) + center_[2];
```

Comment:
Why not `x, y,` and then `z`? This cause be to start searching for some sort of optimization. Funny enough, there could be two (this brings you down to zero extra allocations):
```
ReturnType z = sqrt(square(xbar) + square(ybar));
ReturnType y = Endcap_detail::sin_ax_over_x(z, theta_max_); 
y *= radius_;
ReturnType x = y * xbar + center_[0];
ReturnType y = y * ybar + center_[1];
z = radius_ * cos(z * theta_max_) + center_[2];
```